R/vec-coerce.R

```diff
@@ -121,9 +122,56 @@ def_method2("factor", "factor",
 )
 
 
-### Default coercion to list()
+### Generic coercions
+
+# `NULL` upcoerces to an empty whichever type.
+# Relies on `vec[0]` semantics.
+def_method2("NULL", whichever(),
+  vec_coerce = function(from, to, ...) {
+    if (is_null(from)) {
+      vec_empty(.dispatched[[whichever()]])
+    } else {
+      from
+    }
+  }
+)
+
+# `NA` is treated generically to upcoerce to whichever type.
+# Relies on `vec[0]` semantics and `[<-`.
+def_method2("logical", whichever(),
+  vec_coerce = function(from, to, ...) {
+    # Dispatch to default method if logical is not `NA`
+    if (!identical(.dispatched$logical, NA)) {
```

Comment:
I think a vector of only logical `NA` should have a special type so that this gets handled at the method dispatch level.

Reply:
oh that makes sense, currently I only treat scalar `NA` specially but it's more consistent to treat vectors of `NA` the same. Finding out the type should be fast in the common case of non-NA logical vectors.

What should be the name of this type? `na_vec`?

Reply:
Maybe a `NA` vector should be our whichever type?